### PR TITLE
Datum access

### PR DESCRIPTION
**Pull request by lrennels**
#677 

Currently we have implemented the methods and constructors needed to access parameters and variables in the top level composite components of a model with a call like

```
m[:top, :par_name
```
next steps are to add more comprehensive testing and think about corner cases like renaming.  also, do we want to be able to use this syntax for lower levels i.e. if `:B` is a child composite component of `:top` but not in the `comps_dict` of `m.mi`, should we be able to do 
```
m[:B, :par name]
```

### CODE REVIEW

File: src/core/build.jl [hunk 1]
```diff
@@ -269,7 +270,27 @@ function _build(comp_def::AbstractCompositeComponentDef,
...
+
+    return CompositeComponentInstance(comps, comp_def, time_bounds, variables, parameters)
+end
+
+# helper functions for to create the variables and parameters NamedTuples for a 
+# CompositeComponentInstance
+function _get_variables(comp_def::AbstractCompositeComponentDef, var_dict::Dict{ComponentPath, Any})
+
+    # TODO (dummy below)
+    NT = NamedTuple(Tuple(:var1, :var2), Tuple(1.0, 2.0))
```
Comment (lrennels) [inline src/core/build.jl:285]: @ckingdon95 advice here would be great!

File: src/core/build.jl [hunk 2]
```diff
@@ -269,7 +270,27 @@ function _build(comp_def::AbstractCompositeComponentDef,
...
+    return NT
+end
+
+function _get_parameters(comp_def::AbstractCompositeComponentDef, par_dict::Dict{Tuple{ComponentPath, Symbol}, Any})
+
+    # TODO (dummy below)
+    NT = NamedTuple(Tuple(:par1, :par2), Tuple(1.0, 2.0))
+    return NT
```
Comment (lrennels) [inline src/core/build.jl:293]: and here!

Comment (ckingdon95) [reply src/core/build.jl:293]: and the only difference for the parameters is that CompositeParameterDef's can have multiple refs (they have a `refs` field instead of a `ref` field). As we discussed, all of them will have the same values at run time, so you can just use the first one `par_def.refs[1]`

File: src/core/build.jl (hunk 1 above)
Comment (ckingdon95) [reply src/core/build.jl:285]: so all this needs to do is loop through the comp_def's namespace (which is a dictionary) and keep everything that is a CompositeVariableDef. Then the symbols in the NameTuple constructor will just be the keys from the namespace, and the values in the second half of the constructor will be the vardef.ref (which gives an UnnamedReference object) from each vardef

```
# pseudo code

var_defs = filter(isa(CompositeVariableDef), com_pdef.namespace)  # this probably doesn't quite work for filtering a dictionary on it's values
symbols = [k for (k, v) in var_defs]
vals = [v.ref for (k, v) in var_defs]
NT = NamedTuple(symbols, vals)

File: src/core/instances.jl [hunk 3]
```diff
@@ -168,10 +168,37 @@ function Base.getindex(obj::AbstractCompositeComponentInstance, comp_name::Symbo
...
+        else
+            error("$datum_name is not a parameter or a variable in component $(ci.comp_path).")
+        end
+    end
+    
+    ref = getproperty(which, datum_name)
+    leaf_comp_name = ref.comp_name
+    leaf_datum_name = ref.datum_name
+    
+    # LFR TODO figure out which path to go down to find the leaf
```
Comment (lrennels) [inline src/core/instances.jl:191]: @ckingdon95 as we discussed, we need to just get the next component and datum name to recursively call on, not the final ones represented in the `UnnamedReference`.  I think we could sort of BFS/DFS down to find the leaf and then return it, but wanted to check in to see if you had ideas about where the info would be saved?

File: src/core/instances.jl [hunk 4]
```diff
@@ -168,10 +168,33 @@ function Base.getindex(obj::AbstractCompositeComponentInstance, comp_name::Symbo
...
+        if datum_name in keys(pars) # could merge with below if made names(NamedTuple) = keys(NamedTuple)
+            which = pars
+        else
+            error("$datum_name is not a parameter or a variable in component $(ci.comp_path).")
+        end
+    end
+    
+    ref = getproperty(which, datum_name)
+    
+    # LFR TODO get the comp_name and datum_name to use ...
```
Comment (lrennels) [inline src/core/instances.jl:189]: @ckingdon95 this is a little bit tricky, because sometimes `UnnamedReference` skips a layer, so the `ref.comp_name` is not held in the `ci.comps_dict` ie. in the test we have currently we have a `CompositeComponentInstance` where parameters are fine:
```
julia> ci.parameters 
(fooA1 = Mimi.UnnamedReference(:A, :foo1), 
fooA2 = Mimi.UnnamedReference(:A, :foo2), 
foo3 = Mimi.UnnamedReference(:B, :foo3),
foo4 = Mimi.UnnamedReference(:B, :foo4), 
par_1_1 = Mimi.UnnamedReference(:A, :par_1_1))
```
but the variable just goes all the way to `:Comp3`
```
julia> ci.variables
(var_3_1 = Mimi.UnnamedReference(:Comp3, :var_3_1),)
```
and thus calling `_get_datum(ref.comp_name, ref.datum_name)` won't work...

File: src/core/instances.jl [hunk 5]
```diff
@@ -168,10 +168,33 @@ function Base.getindex(obj::AbstractCompositeComponentInstance, comp_name::Symbo
...
+    return _get_datum(ci.comps_dict[comp_name, datum_name])
```
Comment (lrennels) [inline src/core/instances.jl:190]: I hoped this could just be `_get_datum(ci.comps_dict[ref.comp_name], ref.datum_name)` but see comment above

File: src/core/build.jl [hunk 6]
```diff
@@ -241,9 +241,10 @@ function _instantiate_params(comp_def::ComponentDef, par_dict::Dict{Tuple{Compon
     return ComponentInstanceParameters(names, types, vals, paths)
 end
 
-function _instantiate_params(comp_def::AbstractCompositeComponentDef, par_dict::Dict{Tuple{ComponentPath, Symbol}, Any})
-    _combine_exported_pars(comp_def)
-end
+# commented out because potentially doing nothing ... the fcn is never called
+# function _instantiate_params(comp_def::AbstractCompositeComponentDef, par_dict::Dict{Tuple{ComponentPath, Symbol}, Any})
+#     _combine_exported_pars(comp_def)
+# end
```
Comment (ckingdon95) [inline src/core/build.jl:247]: Can you delete this function and also the `_combine_exported_pars` function too?

File: test/test_composite.jl [hunk 7]
```diff
@@ -138,14 +138,30 @@ run(m)
...
 @test mi["/top/B/Comp4", :par_4_1] == collect(6.0:6:96.0)
 
-#
+@test m[:top, :fooA1] == 1
+@test m[:top, :var_3_1] == collect(6.0:6:96.0)
+
+# vars
+
 # Test joining external params.
 #
```
Comment (ckingdon95) [inline test/test_composite.jl]: Can you clean these comment lines? I think the # got deleted for the next block, and maybe you added `# vars` not needed anymore?

**Comment by lrennels** [comment]
Hmm so would the idea of not naming them `variable` and `parameter` be that they are difference since they are immutable at this level and just references to lower components as opposed to unique ones?

**Comment by ckingdon95** [comment]
yeah mostly the second point-- just that they don't actually store parameter/variable values, but references to them. minor point, but I feel like David would have a clear opinion on it so we could just revisit it later!
